socrata/socrata_upload.py

```diff
@@ -0,0 +1,337 @@
+    all years and upload via `post` (update rather than overwrite).
+    """
+
+    athena_asset, asset_id, row_identifier = get_asset_info(socrata_asset)
+
+    flag, groups = generate_groups(
+        years=years, by_township=by_township, athena_asset=athena_asset
+    )
+
+    tic = time.perf_counter()
+    count = 0
+
+    if not flag:
+        sql_query = build_query(
+            athena_asset=athena_asset,
+            row_identifier=row_identifier,
+        )
+
+        upload_args = {
+            "asset_id": asset_id,
+            "sql_query": sql_query,
+            "overwrite": overwrite,
+        }
+
+        if overwrite:
+            response = upload("put", **upload_args)
+        else:
+            response = upload("post", **upload_args)
+        print(response.content)
+
+    else:
+        if flag == "years":
+            sql_query = build_query(
+                athena_asset=athena_asset,
+                row_identifier=row_identifier,
+                years=years,
+            )
+
+        if flag == "both":
+            sql_query = build_query(
+                athena_asset=athena_asset,
+                row_identifier=row_identifier,
+                years=years,
+                township=by_township,
+            )
+
+        if overwrite:
+            for item in groups:
+                if flag == "both":
+                    upload_args = {
+                        "asset_id": "asset_id",
+                        "sql_query": "sql_query",
+                        "overwrite": "overwrite",
+                        "year": item[0],
+                        "township": item[1],
+                    }
+                elif flag == "years":
+                    upload_args = {
+                        "asset_id": asset_id,
+                        "sql_query": sql_query,
+                        "overwrite": overwrite,
+                        "year": item,
+                    }
+                if count == 0:
+                    response = upload("put", **upload_args)
+                else:
+                    response = upload("post", **upload_args)
+                print(response.content)
+                count = count + 1
+
+        else:
+            for item in groups:
+                if flag == "both":
+                    upload_args = {
+                        "asset_id": asset_id,
+                        "sql_query": sql_query,
+                        "overwrite": overwrite,
+                        "year": item[0],
+                        "township": item[1],
+                    }
+                elif flag == "years":
+                    upload_args = {
+                        "asset_id": asset_id,
+                        "sql_query": sql_query,
+                        "overwrite": overwrite,
+                        "year": item,
+                    }
+                response = upload("post", **upload_args)
+                print(response.content)
```

Comment:
**[Suggestion, non-blocking]** I wonder if we could consolidate these two conditional branches if we treated `overwrite` as something that only matters for determining whether subsequent requests after the first request should be `PUT` or `POST`? Then we'd only need one block that looks something like this:

```python
for item in groups:
    if flag == "both":
        upload_args = {
            "asset_id": "asset_id",
            "sql_query": "sql_query",
            "overwrite": "overwrite",
            "year": item[0],
            "township": item[1],
        }
    elif flag == "years":
        upload_args = {
            "asset_id": asset_id,
            "sql_query": sql_query,
            "overwrite": overwrite,
            "year": item,
        }
    if count == 0 and overwrite:
        response = upload("put", **upload_args)
    else:
        response = upload("post", **upload_args)
    print(response.content)
    count = count + 1
```

Reply:
Perfect.